### PR TITLE
Add Byte type support

### PR DESCRIPTION
**Pull request by AmosLewis**
[https://github.com/llvm/torch-mlir/issues/1383](https://github.com/llvm/torch-mlir/issues/1383)

### CODE REVIEW

**Comment by AmosLewis** [comment]
The type has been added successfully and fix the type bug for gpt2. 
But I got a test bug don't know how to fix. Not sure how to add the test.
```
(mlir_venv) nod% cmake --build build --target check-torch-mlir
[53/54] Running the torch-mlir regression tests
FAIL: TORCH_MLIR :: Conversion/TorchToLinalg/basic.mlir (1 of 103)
******************** TEST 'TORCH_MLIR :: Conversion/TorchToLinalg/basic.mlir' FAILED ********************
Script:
--
: 'RUN: at line 1';   /home/chi/src/ubuntu20/shark/torch-mlir/build/bin/torch-mlir-opt </home/chi/src/ubuntu20/shark/torch-mlir/test/Conversion/TorchToLinalg/basic.mlir -convert-torch-to-linalg -split-input-file -verify-diagnostics | FileCheck /home/chi/src/ubuntu20/shark/torch-mlir/test/Conversion/TorchToLinalg/basic.mlir
--
Exit Code: 1

Command Output (stderr):
--
within split at <stdin>:96 offset :10:8: error: unexpected error: 'torch_c.from_i64' op operand #0 must be 64-bit signless integer, but got 'i8'
  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[],ui8> -> !torch.int
       ^
within split at <stdin>:109 offset :22:8: error: unexpected error: 'torch_c.from_i64' op operand #0 must be 64-bit signless integer, but got 'i8'
  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[?,?],ui8> -> !torch.int
       ^
/home/chi/src/ubuntu20/shark/torch-mlir/test/Conversion/TorchToLinalg/basic.mlir:98:17: error: CHECK-LABEL: expected string not found in input
// CHECK-LABEL: func.func @torch.aten.Int.Tensor$zero_rank
                ^
<stdin>:47:48: note: scanning from here
 func.func @torch.aten.Int.Tensor$non_zero_rank(%arg0: !torch.vtensor<[?,?],si64>) -> !torch.int {
                                               ^
<stdin>:110:2: note: possible intended match here
 func.func @torch.aten.Bool.Tensor$zero_rank(%arg0: !torch.vtensor<[],i1>) -> !torch.bool {
 ^

Input file: <stdin>
Check file: /home/chi/src/ubuntu20/shark/torch-mlir/test/Conversion/TorchToLinalg/basic.mlir

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            .
            .
            .
           42: } 
           43:  
           44:  
           45: // ----- 
           46: module { 
           47:  func.func @torch.aten.Int.Tensor$non_zero_rank(%arg0: !torch.vtensor<[?,?],si64>) -> !torch.int { 
label:98'0                                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           48:  %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[?,?],si64> -> tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %c0 = arith.constant 0 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  %1 = tensor.dim %0, %c0 : tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  %c1 = arith.constant 1 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  %2 = tensor.dim %0, %c1 : tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
          105: } 
label:98'0     ~~
          106:  
label:98'0     ~
          107:  
label:98'0     ~
          108: // ----- 
label:98'0     ~~~~~~~~~
          109: module { 
label:98'0     ~~~~~~~~~
          110:  func.func @torch.aten.Bool.Tensor$zero_rank(%arg0: !torch.vtensor<[],i1>) -> !torch.bool { 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:98'1      ?                                                                                           possible intended match
          111:  %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[],i1> -> tensor<i1> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          112:  %c1_i64 = arith.constant 1 : i64 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          113:  %c0 = arith.constant 0 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  %1 = tensor.extract %0[] : tensor<i1> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          115:  %2 = torch_c.from_i1 %1 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
>>>>>>

--

********************
********************
Failed Tests (1):
  TORCH_MLIR :: Conversion/TorchToLinalg/basic.mlir

```

File: test/Conversion/TorchToLinalg/basic.mlir [hunk 1]
```diff
@@ -95,6 +95,44 @@ func.func @torch.aten.Int.Tensor$non_zero_rank(%arg0: !torch.vtensor<[?,?],si64>
...
+// CHECK:               return %[[RET]] : !torch.int
+func.func @torch.aten.Int.Tensor$zero_rank(%arg0: !torch.vtensor<[],ui8>) -> !torch.int {
+  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[],ui8> -> !torch.int
+  return %0 : !torch.int
+}
+
+// -----
+
+// CHECK-LABEL:     func.func @torch.aten.Int.Tensor$non_zero_rank
+// CHECK-SAME:          (%[[ARG:.*]]: !torch.vtensor<[?,?],si64>) -> !torch.int {
```
Comment (ramiro050) [inline test/Conversion/TorchToLinalg/basic.mlir:112]: I think the input dtype here should be `ui8`

Comment (AmosLewis) [reply test/Conversion/TorchToLinalg/basic.mlir]: Got it fixed. But still got similar bug:

```
Input was:
<<<<<<
           42: } 
           43:  
           44:  
           45: // ----- 
           46: module { 
           47:  func.func @torch.aten.Int.Tensor$non_zero_rank(%arg0: !torch.vtensor<[?,?],si64>) -> !torch.int { 
label:98'0                                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           48:  %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[?,?],si64> -> tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %c0 = arith.constant 0 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  %1 = tensor.dim %0, %c0 : tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  %c1 = arith.constant 1 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  %2 = tensor.dim %0, %c1 : tensor<?x?xi64> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          105: } 
label:98'0     ~~
          106:  
label:98'0     ~
          107:  
label:98'0     ~
          108: // ----- 
label:98'0     ~~~~~~~~~
          109: module { 
label:98'0     ~~~~~~~~~
          110:  func.func @torch.aten.Bool.Tensor$zero_rank(%arg0: !torch.vtensor<[],i1>) -> !torch.bool { 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:98'1      ?                                                                                           possible intended match
          111:  %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[],i1> -> tensor<i1> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          112:  %c1_i64 = arith.constant 1 : i64 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          113:  %c0 = arith.constant 0 : index 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  %1 = tensor.extract %0[] : tensor<i1> 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          115:  %2 = torch_c.from_i1 %1 
label:98'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

```

**Comment by AmosLewis** [comment]
@vivekkhandelwal1

File: test/Conversion/TorchToLinalg/basic.mlir (hunk 1 above)
Comment (ramiro050) [reply test/Conversion/TorchToLinalg/basic.mlir]: Ah, I think the issue is that the tests have the same name as the tests for `si64`. Can you give them a unique name, such as `func.func @torch.aten.Int.Tensor$non_zero_rank_bool` and check if that fixes it?

**Comment by vivekkhandelwal1** [comment]
Hi @AmosLewis, your tests are failing because the TorchToLinalg conversion is not happening. I tried this out and got the following error:
```
within split at test/Conversion/TorchToLinalg/basic.mlir:96 offset :10:8: error: unexpected error: 'torch_c.from_i64' op operand #0 must be 64-bit signless integer, but got 'i8'
  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[],ui8> -> !torch.int
```

**Comment by vivekkhandelwal1** [comment]
I think the `aten.Int.Tensor` op doesn't support ui8 type.

**Comment by ramiro050** [comment]
> Hi @AmosLewis, your tests are failing because the TorchToLinalg conversion is not happening.

Ah, you're right! The `TorchToLinalg` pattern needs a cast at the end.

@AmosLewis, can you replace this line

https://github.com/llvm/torch-mlir/blob/4ef6e69ed49f309aa70ea40f12bf7488cd8e3434/lib/Conversion/TorchToLinalg/TensorScalarInterop.cpp#L100

with 

```cpp
    Value extract = rewriter.create<tensor::ExtractOp>(loc, input, indices);
    Type resultType =
        this->getTypeConverter()->convertType(op->getResult(0).getType());
    rewriter.replaceOp(
        op, convertScalarToDtype(rewriter, loc, extract, resultType));
```

and see if that fixes it? (note: you will need to update the MLIR checks in the unit tests to match the extra IR created by `convertScalarToDtype`)

**Comment by ramiro050** [comment]
> I think the `aten.Int.Tensor` op doesn't support ui8 type.

From what I can tell, it does support it:

```python
In [6]: torch.ops.aten.Int(torch.tensor(0, dtype=torch.int8))
Out[6]: 0
```

**Comment by AmosLewis** [comment]
> > I think the `aten.Int.Tensor` op doesn't support ui8 type.
> 
> From what I can tell, it does support it:
> 
> ```python
> In [6]: torch.ops.aten.Int(torch.tensor(0, dtype=torch.int8))
> Out[6]: 0
> ```

ui8 is uint8, not int8. I think Vivek is right, uint8 is not supported in aten.Int.Tensor . We might need to add aten.Uint.Tensor and torch.uint8 in torch dialect.

**Comment by AmosLewis** [comment]
> > Hi @AmosLewis, your tests are failing because the TorchToLinalg conversion is not happening.
> 
> Ah, you're right! The `TorchToLinalg` pattern needs a cast at the end.
> 
> @AmosLewis, can you replace this line
> 
> https://github.com/llvm/torch-mlir/blob/4ef6e69ed49f309aa70ea40f12bf7488cd8e3434/lib/Conversion/TorchToLinalg/TensorScalarInterop.cpp#L100
> 
> with
> 
> ```c++
>     Value extract = rewriter.create<tensor::ExtractOp>(loc, input, indices);
>     Type resultType =
>         this->getTypeConverter()->convertType(op->getResult(0).getType());
>     rewriter.replaceOp(
>         op, convertScalarToDtype(rewriter, loc, extract, resultType));
> ```
> 
> and see if that fixes it? (note: you will need to update the MLIR checks in the unit tests to match the extra IR created by `convertScalarToDtype`)

Not fix. But bring new errors.  [tmp.mlir](https://gist.github.com/AmosLewis/aa83cf1f371258fae1d77b12bd31dd28)
```
(mlir_venv) nod% /home/chi/src/ubuntu20/shark/torch-mlir/build/bin/torch-mlir-opt </home/chi/src/ubuntu20/shark/torch-mlir/test/Conversion/TorchToLinalg/tmp.mlir -convert-torch-to-linalg -split-input-file
Debug print: 
Debug print: i64
within split at <stdin>:1 offset :2:8: error: unsupported byte, char or bool type for convertScalarToDtype 'i8'(scalar type) -> 'i64'(dtype)
  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[],ui8> -> !torch.int
       ^
within split at <stdin>:1 offset :2:8: error: failed to legalize operation 'torch.aten.Int.Tensor' marked as erased
  %0 = torch.aten.Int.Tensor %arg0 : !torch.vtensor<[],ui8> -> !torch.int
       ^
within split at <stdin>:1 offset :2:8: note: see current operation: %5 = "torch.aten.Int.Tensor"(%arg0) : (!torch.vtensor<[],ui8>) -> !torch.int
within split at <stdin>:1 offset :3:3: note: found live user of result #0: func.return %3 : !torch.int
  return %0 : !torch.int
  ^

```

**Comment by AmosLewis** [comment]
> !torch.int

**Comment by AmosLewis** [comment]
> From what I can tell, it does support it:

I think it is not aten.Int.Tensor but the !torch.int that does not support uint8, we might need to add a torch.uint8 type in torch dialect? @vivekkhandelwal1

**Comment by ramiro050** [comment]
The error you're seeing is a limitation in the `convertScalarToDtype` function:

https://github.com/llvm/torch-mlir/blob/4ef6e69ed49f309aa70ea40f12bf7488cd8e3434/lib/Conversion/Utils/Utils.cpp#L248

Support needs to be added to handle the `Byte` type

File: python/torch_mlir_e2e_test/test_suite/scalar.py [hunk 2]
```diff
@@ -311,3 +311,42 @@ def forward(self):
...
+        None,
+        ([], torch.uint8, True),
+    ])
+
+    def forward(self, val):
+        return int(val)
+
+@register_test_case(module_factory=lambda: AtenIntTensorByteDtypeModule())
+def AtenIntTensorByteDtypeModule_basic(module, tu: TestUtils):
+    module.forward(tu.randint().to(dtype=torch.uint8))
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/scalar.py:332]: These two tests should have a range of negative and positive inputs to make sure the signedness is being handled correctly

File: test/Conversion/TorchToLinalg/basic.mlir [hunk 3]
```diff
@@ -95,6 +95,38 @@ func.func @torch.aten.Int.Tensor$non_zero_rank(%arg0: !torch.vtensor<[?,?],si64>
 
 // -----
 
+// CHECK-LABEL:     func.func @torch.aten.Int.Tensor$zero_rank$byte_dtype
+// CHECK-SAME:          (%[[ARG:.*]]: !torch.vtensor<[],ui8>) -> !torch.int {
+// CHECK:               %[[I:.*]] = torch_c.to_builtin_tensor %[[ARG]] : !torch.vtensor<[],ui8> -> tensor<i8>
+// CHECK:               %[[C1_I64:.*]] = arith.constant 1 : i64
+// CHECK:               %[[C0:.*]] = arith.constant 0 : index
+// CHECK:               %[[EXTRACT:.*]] = tensor.extract %[[I]][] : tensor<i8>
+// CHECK:               %[[RES:.*]] = arith.extsi %[[EXTRACT]] : i8 to i64
```
Comment (ramiro050) [inline test/Conversion/TorchToLinalg/basic.mlir:104]: This doesn't seem right to me. In the case of `ui8`, the op generated for the extension should be `arith.extui`

File: python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py [hunk 4]
```diff
@@ -21,19 +21,23 @@
...
     "mrf64": np.float64,
     "mri1": np.bool_,
+    "mrui8": np.uint8,
+    "mri8": np.int8,
     "mri32": np.int32,
     "mri64": np.int64
 }
 elemental_type_to_ctype = {
     "i1": ctypes.c_bool,
+    "ui8": ctypes.c_ubyte,
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py:39]: I don't think `ui8` is a type that exists in MLIR

File: lib/Conversion/Utils/Utils.cpp [hunk 5]
```diff
@@ -235,20 +235,24 @@ Value convertScalarToDtype(OpBuilder &b, Location loc, Value scalar,
...
     if (auto integerTy = type.dyn_cast<mlir::IntegerType>()) {
       return integerTy.getWidth() == 8;
     }
     return false;
   };
 
-  if (isByteOrChar(scalarType) || isByteOrChar(dtype)) {
+  // We only support conversion from Byte or Char scalarType not to Byte or Char
+  // dtype.
+  if (isByteOrChar(dtype)) {
```
Comment (ramiro050) [inline lib/Conversion/Utils/Utils.cpp:250]: This isn't the right solution. The code below makes the assumption that `ui8` will not make it past this point:

https://github.com/llvm/torch-mlir/blob/a329356fbfce6f17ecf6899a32a7d623812bb8b4/lib/Conversion/Utils/Utils.cpp#L302-L303

Also see TODO: 

https://github.com/llvm/torch-mlir/blob/a329356fbfce6f17ecf6899a32a7d623812bb8b4/lib/Conversion/Utils/Utils.cpp#L238

**Review by ramiro050 — CHANGES_REQUESTED** [review]
I'm not really sure how the handling of `ui8` should be done, since by the time we reach the `linalg` level, the information of whether the input was signed or unsigned is no longer in the IR, so casting the result to the right type is tricky. @silvasean, any thoughts on this?

File: python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py (hunk 4 above)
Comment (silvasean) [reply python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py:39]: Yes, at this level, ui8 should not exist.

**Comment by silvasean** [comment]
> I'm not really sure how the handling of `ui8` should be done, since by the time we reach the `linalg` level, the information of whether the input was signed or unsigned is no longer in the IR, so casting the result to the right type is tricky. @silvasean, any thoughts on this?

We will need to pass the unconverted type around to distinguish these cases.

**Comment by vivekkhandelwal1** [comment]
> > I'm not really sure how the handling of `ui8` should be done, since by the time we reach the `linalg` level, the information of whether the input was signed or unsigned is no longer in the IR, so casting the result to the right type is tricky. @silvasean, any thoughts on this?
> 
> We will need to pass the unconverted type around to distinguish these cases.

Does that mean we can't support this patch as of now? @silvasean @ramiro050

**Comment by silvasean** [comment]
> > > I'm not really sure how the handling of `ui8` should be done, since by the time we reach the `linalg` level, the information of whether the input was signed or unsigned is no longer in the IR, so casting the result to the right type is tricky. @silvasean, any thoughts on this?
> > 
> > 
> > We will need to pass the unconverted type around to distinguish these cases.
> 
> Does that mean we can't support this patch as of now? @silvasean @ramiro050

I think it is worth a shot to pass around the unconverted type -- I don't think it will be that hard. If it does get too difficult, Ramiro or I can work on generalizing the infrastructure to enable this patch.

**Comment by vivekkhandelwal1** [comment]
Hi @ramiro050 @silvasean, I think the latest changes made by me address the issues related to this PR.

**Comment by ramiro050** [comment]
@silvasean my main concern was more regarding the case of a function that returns a tensor of dtype `ui8`. Since this information gets dropped when we get to linalg, the backend running the linalg module will not be able to return the right tensor type. For example, using this patch on a simple identity function running on the refbackend, I get the following:

```
// OuputType: torch
module attributes {torch.debug_module_name = "MyModule"} {
  func.func @forward(%arg0: !torch.vtensor<[?,?],ui8>) -> !torch.vtensor<[?,?],ui8> {
    return %arg0 : !torch.vtensor<[?,?],ui8>
  }
}

// OuputType: linalg-on-tensors
module attributes {torch.debug_module_name = "MyModule"} {
  func.func @forward(%arg0: tensor<?x?xi8>) -> tensor<?x?xi8> {
    return %arg0 : tensor<?x?xi8>
  }
}

Running Compiled Graph
Expected output:
tensor([[  4,   6, 248],
        [247,   9,   8]], dtype=torch.uint8)
Output from compiled MLIR:
[[ 4  6 -8]
 [-9  9  8]]
```

The actual underlying data is the same, but because the assigned dtype is `i8`, if you use the result tensor on other computations, things can go wrong.

**Review by ramiro050 — CHANGES_REQUESTED** [review]

**Comment by silvasean** [comment]
In our current system design there is no way around this ambiguity issue. The workaround would be that for tests that need to check uint8 that they cast to a supported data type when returning from the test.

We could hack around this by preserving the signed/unsigned type on the calling convention boundary, and then have [MungeCallingConventions](https://github.com/llvm/torch-mlir/blob/53e76b8ab606b3d6c3b66e2f970e95d116723375/lib/RefBackend/RefBackend.cpp#L185) recognize that and do the right thing, but that's pretty bad from a layering perspective (implications on the linalg-on-tensors backend contract and other stuff). I think this information generally needs to be kept in a side data structure / attributes, but I haven't thought through the implications of the design space.

**Comment by vivekkhandelwal1** [comment]
Hi @ramiro050, is this PR good to merge?

File: lib/Conversion/Utils/Utils.cpp [hunk 6]
```diff
@@ -229,26 +229,29 @@ SmallVector<Value> getTypeConvertedValues(OpBuilder &b, Location loc,
...
-                           Type dtype) {
+Value convertScalarToDtype(OpBuilder &b, Location loc, Value scalar, Type dtype,
+                           llvm::Optional<Type> srcOriginalDtype) {
   Type scalarType = scalar.getType();
   if (scalarType == dtype)
     return scalar;
 
-  // TODO: For the byte(ui8) or char(i8) case, we need the unconverted dtype to
-  // be able to know if we need signed or unsigned conversion.
+  // TODO: For the conversion to byte(ui8) or char(i8) case, we need the
```
Comment (ramiro050) [inline lib/Conversion/Utils/Utils.cpp:238]: This TODO is no longer needed, since you're passing the original dtype now

**Review by ramiro050 — APPROVED** [review]
just a small nit

**Pull request merged**